docs/code/MlNetCookBook.md

```diff
@@ -0,0 +1,985 @@
+
+However, we also offer a selection of elementary operations that let you customize your NLP processing. Here's the example below where we use them.
+
+Wikipedia detox dataset:
+```
+Sentiment   SentimentText
+1	Stop trolling, zapatancas, calling me a liar merely demonstartes that you arer Zapatancas. You may choose to chase every legitimate editor from this site and ignore me but I am an editor with a record that isnt 99% trolling and therefore my wishes are not to be completely ignored by a sockpuppet like yourself. The consensus is overwhelmingly against you and your trollin g lover Zapatancas,  
+1	::::: Why are you threatening me? I'm not being disruptive, its you who is being disruptive.   
+0	" *::Your POV and propaganda pushing is dully noted. However listing interesting facts in a netral and unacusitory tone is not POV. You seem to be confusing Censorship with POV monitoring. I see nothing POV expressed in the listing of intersting facts. If you want to contribute more facts or edit wording of the cited fact to make them sound more netral then go ahead. No need to CENSOR interesting factual information. "
+0	::::::::This is a gross exaggeration. Nobody is setting a kangaroo court. There was a simple addition concerning the airline. It is the only one disputed here.   
+```
+
+```csharp
+// Create a new environment for ML.NET operations. It can be used for exception tracking and logging, 
+// as well as the source of randomness.
+var env = new LocalEnvironment();
+
+// Define the reader: specify the data columns and where to find them in the text file.
+var reader = TextLoader.CreateReader(env, ctx => (
+        IsToxic: ctx.LoadBool(0),
+        Message: ctx.LoadText(1)
+    ), hasHeader: true);
+
+// Read the data.
+var data = reader.Read(new MultiFileSource(dataPath));
+
+// Inspect the message texts that are read from the file.
+var messageTexts = data.GetColumn(x => x.Message).Take(20).ToArray();
+
+// Apply various kinds of text operations supported by ML.NET.
+var learningPipeline = reader.MakeNewEstimator()
+    .Append(r => (
+        // One-stop shop to run the full text featurization.
+        TextFeatures: r.Message.FeaturizeText(),
+
+        // NLP pipeline 1: bag of words.
+        BagOfWords: r.Message.NormalizeText().ToBagofWords(),
+
+        // NLP pipeline 2: bag of bigrams, using hashes instead of dictionary indices.
+        BagOfBigrams: r.Message.NormalizeText().ToBagofHashedWords(ngramLength: 2, allLengths: false),
+
+        // NLP pipeline 3: bag of tri-character sequences with TF-IDF weighting.
+        BagOfTrichar: r.Message.TokenizeIntoCharacters().ToNgrams(ngramLength: 3, weighting: NgramTransform.WeightingCriteria.TfIdf),
+
+        // NLP pipeline 4: word embeddings.
+        Embeddings: r.Message.NormalizeText().TokenizeText().WordEmbeddings(WordEmbeddingsTransform.PretrainedModelKind.GloVeTwitter25D)
```

Comment:
We should explain why the `NormalizeText()` is very important before the `WordEmbedding()`: Most pre-trained word embedding models contain only lower-case text, by default the NormalizeText lowercases, removes diacritics, though keeps punctuation and numbers. In summary, the user has to ensure their output tokens match their word embedding.

Btw, in the example, we likely want `NormalizeText(keepPunctuations = false)` to due to our poor tokenizer not breaking on punctuation, or use `{ '.', '?', '"', '\'', '!', '-', ';', '(', ')', '_', '|', '#', ... }` as the separators in  `TokenizeText()`.